### PR TITLE
fix(pr): pr table for digests

### LINKED ISSUE #3391: Support pinning docker digests if no tag provided
**What would you like Renovate to be able to do?**
Currently, if I use a digest in something like `.circleci/config.yaml`, I don't get a pin request if I don't provide a tag. So if I do `docker.com/circleci/node`, renovate will ignore it. If I give it a specified version, only then will it attempt to start versioning it and keeping it up to date.

I believe that the docker digest pinning should act the same way NPM does, where in we attempt to pin all the digests on first onboard to renovate, then update them as necessary

**Describe the solution you'd like**
If no tag is provided, determine what `latest` tag maps to through some combination of semver and sha matching and then update the docker digest to have the version. From there, issue PRs when new versions are tagged on docker. Ideally, this option would be a config option, in the event it is not something everyone wants by default.

**Describe alternatives you've considered**
As of today, I've considered generating projects with specific versions so that renovate will at least handle them to an extent. However, I don't know that renovate currently handles major versioning in docker digests, so that is something I am still playing around with.

**Additional context**
I am working in an enterprise, and the reason I want this is so that my downstream consumers using renovate can get a PR to vet the next docker container doesn't indirectly break their builds on some edge case I didn't test.

### PR DESCRIPTION
**Pull request by hasanwhitesource**
## Changes

Added `isDigest` field to LookupUpdate interface.

## Context
Closes #3391 
For the case when `pinDigests` is set to true when scanning a Dockerfile it turns out the `isDigest` is undefined here : 
https://github.com/renovatebot/renovate/blob/15c393f79115c0a5fe6429396c5a9e51127ffc67/lib/workers/repository/updates/generate.ts#L118
Here is a link to [before](https://github.com/renovate-reproductions/3391/pull/1) and [after](https://github.com/hasanwhitesource/DockerDigestRepro/pull/1) to a testing repo.
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Technically `isDigest` is reused today for a digest *update* and not digest pinning. However, we have a problem that we haven't distinguished between pinning *versions* versus pinning *digests*. I think for this PR, `isPin` is more important than `isDigest`

**Comment by hasanwhitesource** [comment]
Also working on getting rid of : 
image

**Comment by rarkins** [comment]
Remember we still want to snow before and after digests when updateType=digest

**Review by viceice — COMMENTED** [review]
Needs test fixes. 😉

**Comment by hasanwhitesource** [comment]
@viceice @rarkins Oh yes also needs some test fixing for expected `updateType`

**Comment by rarkins** [comment]
@hasanwhitesource my concern is that you're fixing a cosmetic problem with a functional change of `updateType` when pinning digests.

File: lib/workers/repository/updates/generate.ts [hunk 1]
```diff
@@ -115,7 +115,7 @@ export function generateBranchConfig(
         upgrade.newDigestShort ||
         upgrade.newDigest.replace('sha256:', '').substring(0, 7);
     }
-    if (upgrade.isDigest) {
+    if (upgrade.isPin) {
```
Comment (hasanwhitesource) [inline lib/workers/repository/updates/generate.ts:118]: @rarkins @viceice This here made me do the functional change.
It was `if(upgrade.isDigest)` before and its undefined when the user uses `pinDigests` flag.
It was setting the change in the table to the tag instead of the digests.

Comment (hasanwhitesource) [reply lib/workers/repository/updates/generate.ts:118]: @rarkins @viceice Another possibility to fix this without changing much is to set the updateType back to `digest`. And then here we change the checking to `upgrade.updateType`.

Comment (rarkins) [reply lib/workers/repository/updates/generate.ts:118]: Yes, would it work to check for `if (upgrade.updateType === 'digest' and !upgrade.currentDigest)` ?

Another possibility is introducing an updateType of `pin-digest`, which maybe should have already existed.

Comment (hasanwhitesource) [reply lib/workers/repository/updates/generate.ts:118]: Yes that would work. But lets do the `pin-digest` type ? sounds more natural in my opinion.

Comment (rarkins) [reply lib/workers/repository/updates/generate.ts:118]: I think it's more correct.. but I'm not sure how much cascading changes it would cause. For example we still want it to be part of any "Pin dependencies" PR

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/updates/generate.ts [hunk 2]
```diff
@@ -115,7 +115,7 @@ export function generateBranchConfig(
         upgrade.newDigestShort ||
         upgrade.newDigest.replace('sha256:', '').substring(0, 7);
     }
-    if (upgrade.isDigest) {
+    if (upgrade.updateType === 'pinDigest' && !upgrade.currentDigest) {
```
[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/workers/repository/updates/generate.ts:118]:
```suggestion
    if (upgrade.updateType === 'pinDigest') {
```

File: lib/util/merge-confidence/index.spec.ts [hunk 3]
```diff
@@ -81,7 +81,17 @@ describe('util/merge-confidence/index', () => {
...
+    it('returns neutral if pinning digest', async () => {
+      expect(
+        await getMergeConfidenceLevel(
+          'npm',
+          'renovate',
+          '25.0.1',
+          '25.0.1',
+          'pinDigest'
+        )
+      ).toBe('neutral');
```
Comment (hasanwhitesource) [inline lib/util/merge-confidence/index.spec.ts:93]:
```suggestion
      ).toBe('high');
```

File: lib/workers/repository/updates/generate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/updates/generate.ts:118]: But what about digest updates? They still exist and are more common. I don't see how you can remove the "if upgrade.isDigest" logic without it changing behavior for digest->digest updates

[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/workers/repository/updates/generate.ts:118]: I see I need to check this case.
I thought this if condition was causing the bug but I think now that we should do an or condition here for the pin digests case ,cause when it was set we were pinning tags.

[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/workers/repository/updates/generate.ts:118]:
```suggestion
    if (upgrade.isDigest || upgrade.updateType === 'pinDigest') {
```
The tests that I've edited were meant for when the `pinDigests` config flag was set to true.
In the case when `pinDigests` was set to true the `upgrade.isDigest` in that if was undefined and the values for the change column were being set to the tag.
therefore I added new update type `pinDigest` which I've added here in case where where pinDigest is set : 
https://github.com/renovatebot/renovate/blob/426dcd20be2a47c9d083de4158da910b98753fd2/lib/workers/repository/process/lookup/index.ts#L337
That sound right ?

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/updates/generate.ts:118]: Yes, I think that adding updateType=pinDigest is logically correct. But you need to make sure you retain backwards compatibility:
- pin and digest updateTypes should behave exactly as before
- pinDigest and pin updateTypes should remain together in a "Pin dependencies" PR

[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/workers/repository/updates/generate.ts:118]: @rarkins Would you please elaborate more on the second bullet that you've mentioned ?
I thought this was only a cosmetic fix. In the case where `pinDigests` was set we were not setting the `upgrade.displayFrom` and `upgrade.displayTo` to the digests, that wasn't the issue ?.

File: lib/util/merge-confidence/index.spec.ts [hunk 4]
```diff
@@ -81,7 +81,17 @@ describe('util/merge-confidence/index', () => {
         )
       ).toBe('high');
     });
-
+    it('returns neutral if pinning digest', async () => {
```
Comment (rarkins) [inline lib/util/merge-confidence/index.spec.ts:84]:
```suggestion
    it('returns high if pinning digest', async () => {
```

File: lib/workers/repository/updates/generate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/updates/generate.ts:118]: Today, both version pins and digest pins are grouped together into the same PR. Your changes need to maintain that backwards compatibility. You could check that by having a repo with one pinned version (e.g. npm `^1.0.0` -> `1.0.1`) and one pinned digest and verifying that they are grouped together like before

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by hasanwhitesource** [comment]
@rarkins I went with the minimal change needed to fix this.
Here is a test repo : https://github.com/hasanwhitesource/DockerDigestRepro/pull/11#issue-1153243402

**Comment by hasanwhitesource** [comment]
@rarkins Is this solution suitable ?

**Comment by rarkins** [comment]
> @rarkins Is this solution suitable ?

What does it look like?

**Comment by rarkins** [comment]
Tested myself.

Before (main branch):
image

After (this PR):
image

Therefore it looks good

**Review by rarkins — APPROVED** [review]
